packages/angular_devkit/build_angular/src/karma/index.ts

```diff
@@ -72,105 +72,114 @@ export function execute(
+
+      if (options.watch !== undefined) {
+        karmaOptions.singleRun = !options.watch;
+      }
+
+      // Convert browsers from a string to an array
+      if (options.browsers) {
+        karmaOptions.browsers = options.browsers.split(',');
+      }
+
+      if (options.reporters) {
+        // Split along commas to make it more natural, and remove empty strings.
+        const reporters = options.reporters
+          .reduce<string[]>((acc, curr) => acc.concat(curr.split(',')), [])
+          .filter(x => !!x);
+
+        if (reporters.length > 0) {
+          karmaOptions.reporters = reporters;
+        }
+      }
+
+      // prepend special webpack loader that will transform test.ts
+      if (options.include && options.include.length > 0) {
+        const mainFilePath = getSystemPath(
+          join(normalize(context.workspaceRoot), options.main),
+        );
+        const files = findTests(options.include, dirname(mainFilePath), context.workspaceRoot);
+        // early exit, no reason to start karma
+        if (!files.length) {
+          throw new Error(
+            `Specified patterns: "${options.include.join(', ')}" did not match any spec files.`,
           );
-          // karma typings incorrectly define start's return value as void
-          // tslint:disable-next-line:no-use-of-empty-return-value
-          const karmaStart = (karmaServer.start() as unknown) as Promise<void>;
-
-          // Cleanup, signal Karma to exit.
-          return () => {
-            // Karma only has the `stop` method start with 3.1.1, so we must defensively check.
-            const karmaServerWithStop = (karmaServer as unknown) as { stop: () => Promise<void> };
-            if (typeof karmaServerWithStop.stop === 'function') {
-              return karmaStart.then(() => karmaServerWithStop.stop());
-            }
-          };
-        }),
-    ),
+        }
+
+        // Get the rules and ensure the Webpack configuration is setup properly
+        const rules = webpackConfig.module?.rules || [];
+        if (!webpackConfig.module) {
+          webpackConfig.module = { rules };
+        } else if (!webpackConfig.module.rules) {
+          webpackConfig.module.rules = rules;
+        }
+
+        rules.unshift({
+          test: mainFilePath,
+          use: {
+            // cannot be a simple path as it differs between environments
+            loader: SingleTestTransformLoader,
+            options: {
+              files,
+              logger: context.logger,
+            },
+          },
+        });
+      }
+
+      karmaOptions.buildWebpack = {
+        options,
+        webpackConfig,
+        // Workaround for https://github.com/karma-runner/karma/issues/3154
+        // When this workaround is removed, user projects need to be updated to use a Karma
+        // version that has a fix for this issue.
+        toJSON: () => { },
+        logger: context.logger,
+      };
+
+      // @types/karma don't include the last parameter.
```

Comment:
FYI I'm working on the fix for this and other issues in @types/karma. Should have a PR up soon.

Reply:
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/52286